Multi-window: window metadata (physical screen, position, dimension) persistence

zxq82lm opened the pull request:
## PR objective
The purpose of this PR is to address item 4 from issue [#290](https://github.com/elodin-sys/elodin/issues/290):
- 4. Extend KDL save/load to persist per-window metadata: physical screen, window position, window size (x, y), and fullscreen state.

## Definitions
- Monitor: the Elodin Monitor object.
- Screen: a physical display.
- Window: a graphical window (here, the main or a secondary one).

## Drone example
```  
        window path="examples/drone/rate-control-panel.kdl" screen=2 {
           rect 5 5 40 40  
        }
        window path="examples/drone/motor-panel.kdl" screen=0 {
           rect 0 0 100 100
        }
        window screen=1 {  
           rect 0 0 50 50    
        }
```   

**Explanations**:
```
- rect 5 5 40 40   # rect <left %> <top %> <width %> <height %>
- rect 0 0 100 100   # full screen
- window screen=1   # no path argument => main window
```

## Potential test interferences
> ⚠️ **Potential test interference**: if the target physical screen already hosts an application (Terminal, Chrome, etc.) in full screen, the window manager switches to exclusive mode (macOS creates a dedicated Space, some X11 WMs maximize as well) and forces our windows to cover 100 % of the screen, even when the KDL specifies a non-fullscreen rect. As long as the user hasn’t exited that full screen, it looks like the layout is ignored. Applies to both macOS and Linux/X11.

>On macOS only: sometimes the OS has already finished leaving fullscreen by the time we call set_fullscreen(None), and AppKit logs a message like:
> ```2025-11-08 16:43:26.588 elodin[15943:3269063] not in fullscreen state```
> This log is harmless and can be safely ignored.

zxq82lm committed:
Add KDL-argument support for screenIdx

zxq82lm committed:
Renaming KDL screenIdx > screen_index + opening window on the right screen (beginning)

zxq82lm committed:
Display a window on the indicated screen (including on reload)

zxq82lm committed:
Detect and spawn screen_index change (load or reload) + logs INFO

zxq82lm committed:
Have the window on the screen specified in the KDL (continued)

zxq82lm committed:
load/reload screen_index working

zxq82lm committed:
Save kdl with screen_index + no reload on save

zxq82lm committed:
Right association (screen, window) before saving

zxq82lm committed:
Add KDL-rect-on-window support

zxq82lm committed:
window > rect support (continued) + cargo clippy

zxq82lm committed:
screen_index > rect (continued) + clippy

zxq82lm committed:
Consolidate the NeedScreen / NeedRect / Idle state machine: only check screen_index

zxq82lm committed:
stabilize secondary window screen/rect handling

zxq82lm committed:
kdl: screen_index tag becomes simply screen

zxq82lm committed:
Stabilize secondary window relayout before handling main window layout

zxq82lm committed:
Add main window screen/rect support to KDL layouts

zxq82lm committed:
Add main-window KDL persistence, layout capture, and safety fallbacks

zxq82lm committed:
Stabilize window layouts, avoid fullscreen loops, persist primary window KDL + fix: ambiguity screen (physical)/monitor (Elodin concept) trough the code

zxq82lm committed:
Fix: handle fullscreen exit. In relation with Ubuntu testing

zxq82lm committed:
Stabilize multi-window layouts: separate mac/Linux (LINUX_MULTI_WINDOW) fullscreen flow, drop persisted titles

zxq82lm committed:
fix: tests::test_save_archive_native_blocks_writes (no link with this branch)

zxq82lm committed:
cargo fmt

zxq82lm committed:
Lay groundwork for Linux multi-window layout flow (macOS path unchanged)

zxq82lm committed:
Linux multi-window flow: wait for fullscreen exit, force windowed rect apply

zxq82lm committed:
Linux: throttle fullscreen waits to keep reload responsive

zxq82lm committed:
Linux: Capture window rects before save (fallback to Bevy Window data) + Clippy

zxq82lm committed:
Linux: persistence of main window layout

zxq82lm committed:
Linux: fix:Save screen/rect KDL; fix:reduce latencies

zxq82lm committed:
cargo fmt

zxq82lm committed:
Clippy

zxq82lm committed:
Linux: Improve multi-window handling: non-blocking fullscreen exit + reliable rect capture

zxq82lm committed:
Linux: few issues fixing according logs

zxq82lm committed:
cargo fmt

zxq82lm committed:
Linux: fix/Prevent hot-loop screen polls and blocking waits in multi-window relayout

zxq82lm committed:
fmt + clippy + default drone  sim.py

zxq82lm committed:
Force XWayland on Linux Wayland and record reliable window rects

zxq82lm committed:
Linux: small fix (Linux compiling)

zxq82lm committed:
Loading KDL (screen, rect) strict separation between macOS & Linux/X11

zxq82lm committed:
Linux/X11: improvements loading rect=...

zxq82lm committed:
Multi-window: fix conflicts, rebasing + no more Bevy/Wayland stuff in Cargo.toml

zxq82lm committed:
env DISPLAY:=0 in the nixhook

zxq82lm committed:
ui: timestamp fullscreen-exit requests so layout timeouts work

zxq82lm committed:
ui: keep KDL-defined window rects from being overwritten

zxq82lm committed:
elodin cli: drop unused Wayland fallback helper (back to main version)

zxq82lm committed:
examples/drone/sim.py: main version

anderspitman reviewed (commented):
This is quite a large amount of code to review. Naively looking at the desired feature, I would hope it ended up being a relatively straight forward modification to our ser/de logic. Obviously it turned out trickier than that.

@zxq82lm would you mind writing (or maybe you already have and I missed it) a high level overview of the issues you ran into and your approach to solving them? That would greatly help me fit everything into place while reviewing.

On another note, some of these files are getting quite large. For example, `libs/elodin-editor/src/ui/mod.rs` has nearly quadrupled in size in the last couple months. Would it be possible to break some of this out into smaller files? Ideally, if code can be encapsulated into individual files with self-contained, pluggable functionality I think it would be easier to review and maintain.

@shanecelis do we already have any sort of convention on file size? I don't tend to be too strict with myself about it, but I've found ~150-300 lines to be ideal, <500 pretty good, and <1000 not too bad.

zxq82lm commented:
> Is it happening just with i3, or also the default Ubuntu window manager?

Reproduced with the default Ubuntu window manager (the fix under testing).

zxq82lm committed:
fix: prevent infinite loop save/load Linux

zxq82lm committed:
Lock the Bevy runtime to X11 via the Nix bootstrap (force X11/XWayland in the Nix shell)

zxq82lm committed:
Clippy

zxq82lm committed:
Multi-window: Optimize performance

zxq82lm committed:
fmt

zxq82lm committed:
fix: restore multi-screen rect locking on macOS (Linux testing pending)

zxq82lm committed:
fmt

zxq82lm committed:
Tiny change in logs + test Linux Ok

zxq82lm committed:
fix: close button on secondary window

zxq82lm committed:
Viewport ok from secondary window

zxq82lm committed:
fix: panic when dragging secondary window with viewport

zxq82lm committed:
Remove window lock system; only write KDL-window on explicit save schematic

zxq82lm committed:
Remove custom titlebar widget usage

zxq82lm committed:
cargo fmt + padding native toolbar on macos

zxq82lm committed:
fix: avoid unused mut on macOS padding setup

anderspitman commented on libs/elodin-editor/src/ui/mod.rs:
```diff
@@ -998,13 +2216,79 @@ fn set_secondary_camera_viewport(
+                return None;
+            }
+            if let RenderTarget::Window(window_ref) = &camera.target {
+                return Some((entity, format!("{:?}", window_ref), camera.order));
```
Since we're using it for comparison, I think we might want to use [NormalizedWindowRef](https://docs.rs/bevy/latest/bevy/window/struct.NormalizedWindowRef.html) instead of format! here.

zxq82lm replied:
Thanks for the pointer -> I didn’t know about NormalizedWindowRef. 
Switched to it so window target comparisons are stable without relying on strings.

zxq82lm committed:
Remove secondary window custom header and rely on native title bar only (Linux)

zxq82lm committed:
fix: add macOS titlebar padding to secondary windows

zxq82lm committed:
fix: show native title text on macOS windows

zxq82lm committed:
fix: make macOS titlebar readable and keep native title visible

anderspitman commented on libs/elodin-editor/src/ui/schematic/load.rs:
```diff
@@ -595,14 +629,159 @@ pub fn graph_label(graph: &Graph) -> String {
+    last_loaded: HashMap<PathBuf, SystemTime>,
+}
+
+impl SchematicLiveReloadRx {
```
See above. Do we still need all this or can we simplify our data flow/source of truth logic to avoid needing it?

zxq82lm replied:
It’s the same concern as the guard above (only on Linux): last_loaded is used to filter bursty notify events alongside the guard.

anderspitman reviewed (dismissed):
@zxq82lm only had time for a few flyby comments until next week, but hopefully at least a little useful.

zxq82lm committed:
sim.py

zxq82lm committed:
format

zxq82lm committed:
Multi-window feature ready on Linux & macOS; kick off optimization pass

zxq82lm committed:
Clippy + remove unuseful comments

zxq82lm committed:
Clear naming distinction between screen (physical screen) vs monitor (Elodin)

zxq82lm commented on libs/elodin-editor/src/ui/mod.rs:
```diff
@@ -663,44 +533,1347 @@ fn sync_secondary_windows(
+    true
+}
+
+fn assign_window_to_screen(
```
Good catch! We’ve now standardized it in the code: “monitor” refers to the monitoring/supervision object in Elodin, and “screen” refers to a physical screen. Thank you.

zxq82lm committed:
fix: fullscreen (maximized) support on load kdl (0 0 100 100) on Linux

zxq82lm committed:
Clippy

zxq82lm committed:
Enable Fifo vsync everywhere; expose native close/min/max on secondary windows

zxq82lm commented on libs/elodin-editor/src/ui/mod.rs:
```diff
@@ -663,44 +533,1347 @@ fn sync_secondary_windows(
-                Window {
-                    title,
-                    resolution: WindowResolution::new(640.0, 480.0),
-                    present_mode: PresentMode::AutoVsync,
```
We now force Fifo (guaranteed vsync) on all platforms and re-enable native close/min/max buttons on secondary windows.

anderspitman replied:
Is there a reason we went with FIFO over AutoVsync?

zxq82lm replied:
> Is there a reason we went with FIFO over AutoVsync?

FIFO guarantees a vsync’d swapchain everywhere (required to be supported), so no tearing and consistent timing across platforms. AutoVsync can fall back to immediate/mailbox if the backend or compositor decides, which can reintroduce tearing or varying latency. We picked FIFO to lock in “always vsync, predictable” behavior.

zxq82lm committed:
Use NormalizedWindowRef for reliable window target comparisons

zxq82lm committed:
Unify log level; silence wgpu_core/wgpu_hal to warn; clean secondary window log output

zxq82lm committed:
Make platform-specific flags compile-time

zxq82lm commented on libs/elodin-editor/src/ui/mod.rs:
```diff
@@ -7,38 +8,64 @@ use bevy::{
+const SCREEN_RELAYOUT_MAX_ATTEMPTS: u8 = 5;
+const SCREEN_RELAYOUT_TIMEOUT: Duration = Duration::from_millis(2000);
+const FULLSCREEN_EXIT_CONFIRMATION_TIMEOUT: Duration = Duration::from_millis(500);
+const LINUX_MULTI_WINDOW: bool = cfg!(target_os = "linux");
```
Converted the platform flags to compile-time cfg (Linux vs non-Linux).

zxq82lm commented on libs/elodin-editor/src/ui/mod.rs:
```diff
@@ -7,38 +8,64 @@ use bevy::{
+// Secondary cameras are offset high to avoid clashing with primary/UI cameras.
+const SECONDARY_GRAPH_ORDER_BASE: isize = 1000;
+const SECONDARY_GRAPH_ORDER_STRIDE: isize = 50;
+const MACOS_PRIMARY_ORDER_OFFSET: isize = if cfg!(target_os = "macos") { 1000 } else { 0 };
```
Converted the platform flags to compile-time cfg (Linux vs non-Linux).

zxq82lm committed:
Drop unused pre-applied rect binding and the macOS frame top inset

zxq82lm commented on libs/elodin-editor/src/ui/mod.rs:
```diff
@@ -663,44 +533,1347 @@ fn sync_secondary_windows(
 
+        // Try to pre-size and pre-position the window to its target rect to avoid an
+        // extra resize pass (and the resulting swapchain churn).
+        let (resolution, position, _pre_applied_rect, pre_applied_screen) = if let Some(rect) =
```
The unused pre-applied rect binding is removed.

zxq82lm committed:
Dropped unecessary code

zxq82lm commented on libs/elodin-editor/src/ui/mod.rs:
```diff
@@ -528,28 +378,30 @@ impl RootWidgetSystem for MainLayout<'_, '_> {
 
-        let mut contexts = state_mut.contexts;
-        let images = state_mut.images;
+        let _contexts = state_mut.contexts;
```
This is no longer necessary; I’ve dropped it.

zxq82lm committed:
Stabilize nav gizmo on secondary windows and drop duplicate viewport handling

zxq82lm commented on libs/elodin-editor/src/ui/command_palette/palette_items.rs:
```diff
@@ -859,18 +859,22 @@ pub fn save_schematic() -> PaletteItem {
+         mut live_reload: ResMut<SchematicLiveReloadRx>| {
             match db_config.schematic_path() {
                 Some(path) => {
+                    live_reload.guard_for(Duration::from_millis(2000));
```
When we save, we write both the main KDL and the secondary window KDLs. On Linux, notify emits multiple events during those writes. Without the guard, live reload can trigger mid-write and reload a half-written file, which can lead to inconsistent schematics, UI flicker, and even “re-creating” secondary windows while they’re still being saved. The guard just suppresses notifications long enough for the writes to finish so we don’t reload prematurely.

I’m not entirely convinced this is the best approach; there might be a cleaner one:
- Instead of a timeout, we could use a cross-thread “saving” flag. The reload thread would check a shared Arc<AtomicBool>/mutex before handling each event, and every save path (main + secondary) would use an RAII guard to set/clear it reliably (even on errors). We’d also have to decide whether to drop or buffer events that arrive while the flag is set. That’s all doable, but it adds coordination and more edge cases. Functionally, the current implementation has been stable and satisfactory in my testing so far.

What’s your opinion on these two approaches? @anderspitman @shanecelis @x46085

anderspitman replied:
@x46085 so that would mean changes to the KDL files are ignored for 5 seconds? Are we sure a user wouldn't want to have a workflow where they're quickly switching back and forth between WYSIWYG changes and editing the KDL directly?

@zxq82lm so the problem is that we need to process the updates to all the files as one atomic group, but the events are coming in separately? Assuming I'm understanding that correctly, I think we might be working around a deeper issue, that if solved may allow us to simplify the approach.

I think the root problem is that saving the schematic from the GUI editor triggers multiple file change notifications that we need to handle. But those notifications are superfluous. The changes originated from GUI editor, so there should be no need to process them. The KDL is the source of truth, but if the editor already matches the source of truth, no changes are necessary. Could we just keep the serialized string of each KDL file (or even a hash) in memory, and whenever a notification for that KDL path comes in, we read the file contents, checked if it's changed, and if not ignore the notification?

This is a bit off the cuff so please let me know if I'm missing something.

zxq82lm replied:
> @zxq82lm so the problem is that we need to process the updates to all the files as one atomic group, but the events are coming in separately? Assuming I'm understanding that correctly, I think we might be working around a deeper issue, that if solved may allow us to simplify the approach.
> 
> I think the root problem is that saving the schematic from the GUI editor triggers multiple file change notifications that we need to handle. But those notifications are superfluous. The changes originated from GUI editor, so there should be no need to process them. The KDL is the source of truth, but if the editor already matches the source of truth, no changes are necessary. Could we just keep the serialized string of each KDL file (or even a hash) in memory, and whenever a notification for that KDL path comes in, we read the file contents, checked if it's changed, and if not ignore the notification?
> 
> This is a bit off the cuff so please let me know if I'm missing something.

- Your read is right: we write a bundle (main + secondary KDLs) and events arrive separately, so the guard prevents reloading half-written files during that burst.
- Hash/serialized comparison could cut redundant reloads, but you still must avoid reading/hashing mid-write; a mixed approach is possible (brief silence, then hash-filter GUI saves).
- Limited gain: the guard already covers the risky cases and stays simple. Hash/cache adds I/O/sync complexity and doesn’t remove the need to protect the write phase, so if things are stable the guard is a reasonable compromise.

zxq82lm committed:
Fix imports after rebase (GridPrecision, drop unused Margin)

shanecelis reviewed (changes requested):
This works, but I'm a little concerned with some of the changes. I have added two commits (with permission, of course). 

One of the pain points I saw with this code is that there's a struggle to express essentially an asynchronous operation within a synchronous code block. I see this in `apply_secondary_window_screens` for instance. I have started a refactor to use `bevy_defer`, which provides an asynchronous API, which I think will help this block of code a lot.

I have not gone through and removed the fields from state that are no longer necessary. It compiles but there is a `todo!` that remains to be resolved.

Finally, there is a fair amount of duplication between the Primary and Secondary windows implementation. For instance there is another system `apply_primary_window_layout` that has the same kind of issues addressed in the refactor commit I mentioned above. It would be lovely to try and reduce some of this code and struct duplication. I wonder if an enum that was `enum EWindow { Primary { ... }, Secondary { ... } }` might help. We had two enums that had the same variants but one was named `SecondaryWindowRelayoutPhase`. I'm happy to discuss it further on discord as you prefer.

I'd encourage you to try leveraging `bevy_defer`. It's a very useful library.

shanecelis committed:
Multi window refactor (#331)

* refactor: Prefer fields to bare accessory methods.

* refactor: Use bevy_defer for async operations.

* feat: async works.

* refactor: Let's use one `WindowId` for all windows.

* feat: Add wait_for_window_to_change_screens() async.

* refactor: Add wait_for_window_to_change_screens().

Remove awaiting_screen_confirmation field.

* refactor: Make relayout an event.

* refactor: Prefer one-off system to ignore flags.

* refactor: Rename to WindowState.

* refactor: Store `WindowState` as component.

* refactor: Use AtomicU32 for auto-increment.

* style: Reformat.

* doc: Add TODOs.

* refactor: Use `Entity` as `WindowState` key.

Instead of `WindowId`, which oftens requires a find().

* feat: Add `WindowState` to primary window.

* refactor: Use `WindowState` for primary window too.

* refactor: Don't need to keep mut var exposed.

* refactor: Use render_layout for all windows.

* refactor: Remove WindowManager and PrimaryWindowLayout.

* refactor: Remove WindowRelayoutPhase, uses event now.

* style: Reformat.

* refactor: Make clippy happy.

* style: Reformat.

* fix: Missing import.

* refactor: Make clippy happy.

zxq82lm committed:
WindowRelayout::UpdateDescriptors fired from (just before) Save Schematic + adjustements

zxq82lm committed:
Disable live window geometry tracking for new secondary windows (from Create Window) + fmt

zxq82lm committed:
Emit UpdateDescriptors before saving; set primary window descriptor path as KDL source of truth

zxq82lm committed:
fmt + Clippy

zxq82lm committed:
“Rely on UpdateDescriptors for capture; remove redundant direct call before save.”

zxq82lm committed:
Polish and clean up multi-window support on Linux

zxq82lm committed:
macos: add settle guard to secondary window rect application

zxq82lm committed:
fmt + clippy

zxq82lm committed:
Fix nav gizmo viewport panic by clamping to target window

zxq82lm committed:
macos: apply window rects in borderless mode to avoid fullscreen Spaces

zxq82lm committed:
macos: rollback settle logic and drop decoration changes to avoid crashes

zxq82lm committed:
Fix query plot pane init and dedupe tree actions

zxq82lm committed:
macos: serialize relayout per window; block rect when fullscreen/off target; propagate settle timeout

zxq82lm committed:
macos: every paths are ok (save schematic, load schematic) but load init

zxq82lm committed:
fmt

zxq82lm committed:
macos/linux: adjustements

zxq82lm committed:
clippy

zxq82lm committed:
clippy

zxq82lm committed:
clippy

zxq82lm committed:
clippy

x46085 reviewed (approved):
Make it so!

shanecelis commented on libs/elodin-editor/src/ui/mod.rs:
```diff
@@ -826,130 +640,950 @@ fn readable_label_color(color: Color) -> Color32 {
+
+#[cfg(not(target_os = "macos"))]
+#[cfg(not(target_os = "macos"))]
+#[cfg(not(target_os = "macos"))]
```
Cfg triple not necessary, right?

shanecelis reviewed (approved):
Working well on macOS!

zxq82lm committed:
fix: cleaning extra cfg
